Use DTS for nRF5 I2C

tkln opened the pull request:
Device tree bindings for the nRF5 I2C controller are added. The i2c_nrf5 driver is modified and SoC dts fixups are added. The dts I2C configuration is enabled for all nRF5 boards

tkln committed:
dts: bindings: i2c: Add bindings for i2c_nrf5

A device tree bindings description is added for nRF5 I2C controller.

Signed-off-by: Aapo Vienamo <aapo.vienamo@iki.fi>

tkln committed:
dts: arm: nrf5: Add i2c nodes

DT nodes for i2c0 and i2c1 are added for the nRF5 SoCs.

Signed-off-by: Aapo Vienamo <aapo.vienamo@iki.fi>

tkln committed:
soc: nrf5: Use i2c configure options generated from dts

SoC dts fixups are added and the i2c_nrf5 driver is modified to use
the values generated from the device tree.

The I2C_*_DEFAULT_CFG and I2C_*_IRQ_PRI options are removed from board
defconfigs. Bitrate and IRQ priority are configured using using the
device tree instead.

HAS_DTS_I2C is selected on all nrf5 boards to prevent generation of
conflicting defines from Kconfig.

Signed-off-by: Aapo Vienamo <aapo.vienamo@iki.fi>

carlescufi commented on dts/bindings/i2c/nordic,nrf5-i2c.yaml:
```diff
@@ -0,0 +1,42 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+---
+title: Atmel SAM Family I2C (TWIHS) node
```
Wrong IC Family

galak reviewed (changes requested):
Commit messages need a body, please fix.

Fix typo in: 'Add bindigs' -> 'Add bindings'

galak commented on dts/arm/nordic/nrf52832.dtsi:
```diff
@@ -44,6 +44,28 @@
+			#size-cells = <0>;
+			reg = <0x40003000 0x1000>;
+			interrupts = <3 1>;
+			default_cfg = <0x04>;
```
We shouldn't have default_cfg, this driver should be updated to use _i2c_map_dt_bitrate with clock-frequency property in dts.  You can see commit 54933b3833 for an example.

galak commented on arch/arm/soc/nordic_nrf5/nrf52/dts.fixup:
```diff
@@ -7,4 +7,16 @@
 
 #define FLASH_DEV_NAME			NRF_NRF52_FLASH_CONTROLLER_4001E000_LABEL
 
+#define CONFIG_I2C_0_BASE_ADDR		((void *)NORDIC_NRF5_I2C_40003000_BASE_ADDRESS_0)
```
We tend to do the void * cast in the driver, not here.

galak commented on arch/arm/soc/nordic_nrf5/nrf51/dts.fixup:
```diff
@@ -7,4 +7,16 @@
 
 #define FLASH_DEV_NAME			NRF_NRF51_FLASH_CONTROLLER_4001E000_LABEL
 
+#define CONFIG_I2C_0_BASE_ADDR		((void *)NORDIC_NRF5_I2C_40003000_BASE_ADDRESS_0)
```
We tend to do the void * cast in the driver, not here.

galak reviewed (changes requested):
Various code cleanups.  Also you don’t enable i2c controller in any board dts files, that seems odd.

tkln committed:
boards: arm: nrf5: Enable i2c dts nodes for boards

The i2c controller nodes are enabled in the board specific dts files
on boards where the controller was enabled in the board Kconfig.

Signed-off-by: Aapo Vienamo <aapo.vienamo@iki.fi>

ydamigos reviewed (commented):
You could also remove the following block from boards' Kconfig.defconfig file.

```
config I2C_x_IRQ_PRI
              default 1
```

tkln commented:
@carlescufi @galak @ydamigos done.

galak commented on boards/arm/96b_carbon_nrf51/96b_carbon_nrf51.dts:
```diff
@@ -23,3 +23,11 @@
 	current-speed = <115200>;
 	status = "ok";
 };
+
```
curious why you enabled any i2c on this board, nothing in the existing code is enabling i2c by default for the board

galak commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -44,6 +45,30 @@
+			reg = <0x40003000 0x1000>;
+			clock-frequency = <I2C_BITRATE_STANDARD>;
+			interrupts = <3 1>;
+			default_cfg = <0x04>;
```
remove default_cfg

galak commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -44,6 +45,30 @@
+			reg = <0x40004000 0x1000>;
+			clock-frequency = <I2C_BITRATE_STANDARD>;
+			interrupts = <4 1>;
+			default_cfg = <0x04>;
```
remove default_cfg

galak commented on dts/arm/nordic/nrf52832.dtsi:
```diff
@@ -44,6 +45,30 @@
+			reg = <0x40003000 0x1000>;
+			clock-frequency = <I2C_BITRATE_STANDARD>;
+			interrupts = <3 1>;
+			default_cfg = <0x04>;
```
remove default_cfg

galak commented on dts/arm/nordic/nrf52832.dtsi:
```diff
@@ -44,6 +45,30 @@
+			reg = <0x40004000 0x1000>;
+			clock-frequency = <I2C_BITRATE_STANDARD>;
+			interrupts = <4 1>;
+			default_cfg = <0x04>;
```
remove default_cfg

galak commented on dts/arm/nordic/nrf52840.dtsi:
```diff
@@ -52,6 +53,30 @@
+			clock-frequency = <I2C_BITRATE_STANDARD>;
+			interrupts = <3 1>;
+			default_cfg = <0x04>;
+			status = "disabled";
```
remove default_cfg

galak commented on dts/arm/nordic/nrf52840.dtsi:
```diff
@@ -52,6 +53,30 @@
+			reg = <0x40004000 0x1000>;
+			clock-frequency = <I2C_BITRATE_STANDARD>;
+			interrupts = <4 1>;
+			default_cfg = <0x04>;
```
remove default_cfg

galak commented on boards/arm/96b_carbon_nrf51/Kconfig.board:
```diff
@@ -6,4 +6,5 @@
 
 config BOARD_96B_CARBON_NRF51
 	bool "96Boards Carbon (nRF51)"
+	select HAS_DTS_I2C
```
In general the HAS_DTS_I2C should be done in drivers/i2c/Kconfig.nrf5

galak reviewed (changes requested):
Various code cleanups commented on.

For the boards, you seem to be enabling i2c for boards that didn’t and aren’t converting the boards that do use I2C over to DTS.

So boards like nrf51_vbluno51, nrf52_vbluno52, and bbc_microbit should update their DTS to enable I2C.  I think all the other boards should be left alone at this time since they don’t seem to enable I2C by default.

ydamigos commented on boards/arm/bbc_microbit/bbc_microbit.dts:
```diff
@@ -22,3 +22,7 @@
 };
+
+&i2c0 {
+	status = "ok";
```
I would propose to add `clock-frequency = <I2C_BITRATE_FAST>;` for all boards in the PR because most users will use 400 kHz bitrate. If you don't, just keep in mind that every user should create a board overlay to override it in application level.

ydamigos reviewed (approved):
LGTM. Just a non-blocking comment.

tkln commented:
@ydamigos Done.

tkln commented:
@carlescufi @anangl Do you see any remaining issues or is this ready for merging?

anangl reviewed (approved):
LGTM.
